### DISCUSSION TITLE
Style and conventions discussion

### DISCUSSION THREAD

**Discussion by taladan** [club-40/odin-pantry#6] [General]
Since we are getting closer to a realized selection of which project we are going to be working on, it occurs to me that we also need to start thinking about what kind of code conventions/style we're wanting to adhere to.  

I do not personally have a preference as I have no exposure to what companies or other FOSS projects are currently using, but wanted to at least get the ball rolling on a discussion as to what others have seen/used/prefer.  

If I'm jumping the gun a bit here, I apologize, but I think that the sooner we get this kind of thing discussed and decided upon, the sooner we'll get to building.

**Comment by Philip-Clark** [comment]
I think TOP recommends ( or at least links ) the Airbnb js style guide. I've looked through that before, and it seems solid.
https://airbnb.io/javascript/

**Reply by RyanMcEntire** [reply]
Lets go with Airbnb for js its widely used and seems reasonable from my limited experience.

**Comment by JoshDevHub** [comment]
This depends a bit on what stack is eventually used. I would assume that whatever project is picked will have JavaScript at the least and probably React as well. I think a good starting point could be using eslint with react/recommended and airbnb extensions.

A potential related decision is whether or not to use TypeScript. I figure more advanced people will be excited by a TypeScript project, but that could come at the cost of people earlier in the curriculum maybe not being able to contribute as easily.

If the project is suited for a Rails backend, then I'm in favor of rubocop (with rails, rspec, and performance extensions) and probably a few alterations to the default rules that I'll advocate for if/when the time comes :sweat_smile:

**Comment by Lofty-Brambles** [comment]
From what it sounds like, sticking to AirBnB with a handful of other important eslint plugins like import should be good for the JS/TS code. As for the ruby code, rubocop is kinda going to be the unanimous answer, so there's that.

As for TS (which's a big question), I'm in the favour of including it. It isn't something entirely new and the only utility it has is a build step and types, and it'll be a learning opportunity for those who are currently at a state in their paths and it helps a ton in jobs as well. It's rare to see most, if any open source codebase give up the blanket benefits of ts and type checking ci.

**Comment by justEstif** [comment]
Contributing conventions should also be discussed. 

While I use conventional commits, I think sticking to everything in TOP's contributing guide will avoid confusions.

**Comment by JaeGif** [comment]
Now that the stack is nearly finalized, I think it's time to begin this discussion again. 

Zustand, TS React, and React Query I don't think will be replaced in the poll based on the current numbers, routing regardless of the implementation is a conventions discussion.

Mostly hungup on naming conventions and general code conventions specific to the stack, especially for queryKey's since the name matters for changing the queries. Will we be using TS in strict exclusively? I think it's achievable, but definitely more difficult for new contributors and people unfamiliar with Typing which should be considered.

I'm not sure if a doc already exists for this so I started a google doc for it. I think using a doc would be easier to keep track of comments and ideas. [doc](https://docs.google.com/document/d/19ls_WMVQ7ZBK05W4AAg7uEz3QZlmpd9ZofRsA6TiHJU/edit?usp=sharing). We can whip it up into a readme when we're finished for contributors.
